default/python/AI/AIFleetMission.py

```diff
@@ -139,79 +135,53 @@ def _get_fleet_order_from_target(self, mission_type, target):
 
     def check_mergers(self, context=""):
         """
-        If possible and reasonable, merge this fleet with others.
+        Merge local fleets with same mission into this fleet.
 
         :param context: Context of the function call for logging purposes
         :type context: str
         """
+        debug("Considering to merge %s", self.__str__())
         if self.type not in MERGEABLE_MISSION_TYPES:
+            debug("Mission type does not allow merging.")
+            return
+
+        if not self.target:
+            debug("Mission has no valid target - do not merge.")
             return
+
         universe = fo.getUniverse()
         empire_id = fo.empireID()
+
         fleet_id = self.fleet.id
         main_fleet = universe.getFleet(fleet_id)
-        system_id = main_fleet.systemID
-        if system_id == INVALID_ID:
-            return  # can't merge fleets in middle of starlane
-        aistate = get_aistate()
-        system_status = aistate.systemStatus[system_id]
-
-        # if a combat mission, and only have final order (so must be at final target), don't try
-        # merging if there is no local threat (it tends to lead to fleet object churn)
-        if self.type in COMBAT_MISSION_TYPES and len(self.orders) == 1 and not system_status.get('totalThreat', 0):
+        main_fleet_system_id = main_fleet.systemID
+        if main_fleet_system_id == INVALID_ID:
+            debug("Can't merge: fleet in middle of starlane")
             return
-        destroyed_list = list(universe.destroyedObjectIDs(empire_id))
+
+        # only merge PROTECT_REGION if there is any threat near target
+        if self.type == MissionType.PROTECT_REGION:
```

Comment:
Consider adding TODO:
```
Logic about cheking if we can merge certain mission types can be moved to MissionType class itself.
```


PS.  It can be easily implemented via inheritance or via mixins.

Mixin approach:

```python

class Base:
    def mergable(self, other):
        return False
    
# has a single method!
class Mergable:
    def mergable(self, other):
        return True
    
class TypaA(Base):
    pass

class TypeB(Base, Mergable):
    pass


class TypeC(Base):
    def mergable(self, other):
        ...
```